src/traits.rs

```diff
@@ -0,0 +1,718 @@
+    /// afford allocation.
+    #[inline]
+    fn cloned<Item>(self) -> Cloned<Self, Item>
+    where
+        Self: for<'a> SequenceItem<'a, Item = &'a Item> + Sized,
+        Item: Clone,
+    {
+        Cloned::new(self)
+    }
+
+    /// Takes a closure and creates a sequence that calls the closure on each element.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use sqnc::traits::*;
+    ///
+    /// let x = [1, 2, 3];
+    /// let y = Sequence::map(x, |v| v + 4);
+    /// assert_eq!(y.get(1), Some(6));
+    /// assert!(y.iter().eq(5..8));
+    /// ```
+    ///
+    /// # Caveat
+    ///
+    /// The map will be called for every element access, even if an element has
+    /// already been accessed. If the map is expensive and elements are going
+    /// to be accessed multiple times, it is probably more efficient to store
+    /// the mapped sequence in a [`Vec`][`std::vec::Vec`], if one can afford
+    /// allocation.
+    #[inline]
+    fn map<B, F>(self, f: F) -> Map<Self, F>
+    where
+        Self: Sized,
+        F: for<'a> Fn(<Self as SequenceItem<'a>>::Item) -> B,
+    {
+        Map::new(self, f)
+    }
+
+    /// Returns the concatenation with another sequence.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use sqnc::traits::*;
+    ///
+    /// let x = [0, 1, 2];
+    /// let y = [3, 4, 5];
+    /// let z = x.concat(y);
+    /// assert!(z.iter().copied().eq(0..6));
+    /// ```
+    #[inline]
+    fn concat<Other>(self, other: Other) -> Concat<Self, Other>
+    where
+        Self: Sized,
+        Other: Sequence + for<'a> SequenceItem<'a, Item = <Self as SequenceItem<'a>>::Item>,
+    {
+        Concat::new(self, other)
+    }
+
+    /// Returns a selection of the sequence or `None` if any index is out of bounds.
+    ///
+    /// The sequence of indices must have [`usize`] as element type.
+    /// [`Sequence::copied()`] can be used to obtain a sequence of owned indices
+    /// given a sequence of references to [`usize`].
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use sqnc::traits::*;
+    ///
+    /// let x = b"cdelst!";
+    /// let i = [4, 2, 3, 2, 0, 5, 2, 1, 6].copied();
+    /// let y = x.select(i).unwrap();
+    /// assert!(y.iter().eq(b"selected!"));
+    ///
+    /// assert_eq!(x.select([4, 8, 0].copied()), None); // Index `8` is out of bounds.
+    /// ```
+    #[inline]
+    fn select<Idx>(self, indices: Idx) -> Option<Select<Self, Idx>>
+    where
+        Self: Sized,
+        Idx: IndexableSequence + IterableSequence + for<'a> SequenceItem<'a, Item = usize>,
+    {
+        Select::new(self, indices)
+    }
+
+    /// Returns a compressed sequence or `None` if the mask and the sequence have different lengths.
+    ///
+    /// The mask must be a sequence of type [`bool`]. [`Sequence::copied()`] can
+    /// be used to obtain a sequence of owned values given a sequence of
+    /// references to [`bool`].
```

Comment:
Maybe it's good to add something along the lines:

Note that, since compress does not allocate any memory, access of individual items requires iteration over the mask. Situation that heavily depend on random access might be better of allocating an index vector and using select instead.

Reply:
I've included a caveat (slightly different from your suggestion) with an example showing how to obtain the index vector.